### DISCUSSION TITLE
Your thoughts on adding (time-domain) TS augmentation as a transformer subclass

### DISCUSSION THREAD

**Discussion by MrPr3ntice** [alan-turing-institute/sktime#1592] [Ideas]
Hi everyone,

as part of our current research, we implemented sktime-compatible (sub-)class for most common TS-augmentation-techniques in time domain (e.g., noise, crop, drift, quantize, jitter, resize, reverse, time warping, permutation, ... see, e.g., [https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0254841](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0254841) ) based on your transfomer base class. Of course, it is sklearn-"pipeline-able" and compatible with both UTS and MTS in the common sktime formats (also for non-tabular MTS).

Question for the community: If there is common interest in this and it aligns with your current development goals, we would look to refactor our existing code to your guidelines, write a STEM on technical details, and wait for your feedback.

Thank you for your response and thoughts, and thanks for your work
Philipp

**Comment by fkiraly** [comment]
I'd say, +100, these would all be great additions!

I assume these transformers act on individual time series as well as panels (collections of time series)?

If it follows the current interface, no STEP is needed.
You can write a pre-implementation design, it may be useful for discussions - but it sounds like you already have code available?

**Reply by MrPr3ntice** [reply]
Thanks for your positive reply! 

Yes, it will be a panel transformer on the first hand and will deal as well with individual ts.

As I am on a training this week, we will try to provide a pre-implementation for discussion purpose by next week. What is the preferred way to provide this (create individual branch, pull-request ...)?

**Reply by fkiraly** [reply]
The "standard" way would be fork-branch-PR.

FYI, there is currently a refactor of transformers ongoing here:
https://github.com/alan-turing-institute/sktime/pull/1365

This may take a little time, days or weeks, but it might help with implementation - the PR contains a transformers extension template (of course for the "post-refactor" state, not pre-refactor).

Would appreciate any feedback on what you think about the refactor, and/or whether you think the extension template helps with implementing. Feel free to post directly in the PR.

**Reply by MrPr3ntice** [reply]
I went through most changes proposed by PR #1365, and it helped a lot to understand the idea behind the not-so-easy inter-compatibility of different data representation in sktime. The proposed implementation of BaseTransformer and the corresponding transformer template are a convincing solution to overcome this issue. I woud prefer to match our code directly to the proposed changes in #1365 as long as it is likely and foreseeable that it is merged to main any soon?

Is there a prefered example for a unit test in [here](https://github.com/alan-turing-institute/sktime/tree/main/sktime/transformations/panel/tests) or even a universal transformer unit test as part of PR #1365?

Regardless of this: Am I correct, that (corresponding to [example](https://github.com/alan-turing-institute/sktime/blob/transformers-refactor/examples/AA_datatypes_and_datasets.ipynb)) there is currently no mtype for series supporting differently sampled variables (e.g. sensor A @ 1Hz and sensor B @ 16.6Hz)? In case of Panel, I often used "nested_univ" and "numpyflat" in the past, but they seem minor / not preferred anymore.

Thank you a lot!

**Reply by fkiraly** [reply]
> I woud prefer to match our code directly to the proposed changes in #1365 as long as it is likely and foreseeable that it is merged to main any soon?

I think it's almost done and almost ready to merge, although as far as I understand, the merge is currently blocked by a little ideological debate among core devs. The question is, what mtype should `transform` return? One side to the discussion thinks it should be the same mtype that is input (where applicable, i.e., if both input/output are `Series` or `Panel`). The other side thinks it should always be the *inner* mtype, i.e., the mtype that the algorithm happens to use internally.

I'm not going to say who thinks what - but given that you have dived into this, do you have an opinion here?

Regarding your implementation: I think it won't change the extender contract (the extension template), since the discussion is about functionality in the base class and does not touch the extension template. So I feel it's probably 90% safe to use the PR state for extension.

> Is there a prefered example for a unit test

Can you kindly rephrase this? What kind of test are you looking for? A test that checks whether a transformer has been implemented correctly?

> Regardless of this: Am I correct, that (corresponding to example) there is currently no mtype for series supporting differently sampled variables

As far as I understand, the following mtypes support this: `nested_univ`; `pd-long`. As far as I understand, `numpyflat` doesn't, since the columns are the product of "variable" with "time index" (so need to be the same for all instances).
They are currently not preferred because core devs and users don't seem to like to work with them, and we have no estimators that deal with unequally sampled data.

However, if you add a lot of things here, e.g., re-sampling, aggregation, etc, we can easily upgrade one of these mtypes to "preferred". I would also like to hear your opinion which mtype you think would be best to work with here, we can also design a new one if it turns out it's easier to use.

**Reply by fkiraly** [reply]
@MrPr3ntice, the transformers PR #1365 is now in!
Happy coding, let us know if you have any questions.
Also, would be interested to hear your opinions above.

**Comment by MrPr3ntice** [comment]
Thanks for your info, we already fetched the updates and are confident to provide a PR showing our structure by the next days. We still have some dtype-checking problems with the new BaseTransformer.fit / transform... maybe it has something to do with our use of `_tags` ... 

In general, we struggle a little with the given division into Series and Panels as a Series is a Panel with a single instance. From this point of view, the division seems unnecessary and Panels should be sufficient, but maybe I am missing something.

Regarding your questions:

>  The question is, what mtype should transform return

I see the benefits of both sides: I love the idea, that the original input mtype remains (is back-transformed) for the output. From my practicle point of view, I see the downside of back-transformation in terms of computational speed and RAM for big datasets. In a perfect world, this may be a decision of the user, either by an init parameter of the transformer or a `_tag = {keep_mtype = True}` with default value `True` i.e. original mtype is returned.

> Can you kindly rephrase this? What kind of test are you looking for? A test that checks whether a transformer has been implemented correctly?

Exactly! A template to adapt an put into `transformations/panel/tests`. For now, we oriented ourselfs towards one random test example from this directory, but maybe there is (also) a preferred structure or universal test for all transformers?

> As far as I understand, the following mtypes support this: `nested_univ`; `pd-long`.

You are right, I mixed up `numpyflat` and `pd-long`. I know the downsides of (multi-)variate TS with unequal length (especially in terms of memory allocation and computational speed) - but in my field, they are common practice as datasets would get too big if all variables are resampled to the same length. `pd-long` is a quite good compromise, but suffers from information overhead and computational expensive indexing (e.g. `data_of_var1 = data.loc[data['variable'] == 'var1', 'value']` to select all data of 'var1') where 2D-nested Dataframes with Series as elements seem quite unintentional by Pandas (and suboptimal for memory allocation of course). Sometimes I wonder what contradicts simple 2D-list sturctures (i.e. [ [pd.Series(), pd.Series(), ...], [pd.Series(), pd.Series(), ...], ...], with instances as first dim and variables as second dim, comparable to `df-list`) as python lists are quite competend in managing memory allocation issues... But I understand that this is a marginal problem and most users will be fine with simply resampling all variable/sensors to the same length if they are not by default.

To sum it up - `nested_univ` works fine for us as `inner_mtype` but feels nevertheless  'wrong' and could easely be kicked out of Pandas in the future, maybe I should make some efforts in a speed test comparison of different solutions (`pd-long` vs. `nested_univ `vs. `list_of_list_of_series`) for this problem in the near future...

I'm looking forward for your feedback on our PR.

**Reply by fkiraly** [reply]
> Thanks for your info, we already fetched the updates and are confident to provide a PR showing our structure by the next days. We still have some dtype-checking problems with the new BaseTransformer.fit / transform...

There's a known bug:
https://github.com/alan-turing-institute/sktime/pull/1663
and I'm currently also writing more comprehensive tests (which will also hopefully result in the testing utility that you are looking for). 

May take a couple days or weeks though, so "testing this in anger" and rapid feedback would be very much appreciated!
We also have a regular "transformers" meeting 4pm UTC on Fridays (discord), if you run into anything, we could do pair programming then.

> In general, we struggle a little with the given division into Series and Panels as a Series is a Panel with a single instance. From this point of view, the division seems unnecessary and Panels should be sufficient, but maybe I am missing something.

Sort of - mathematically, this is correct, but it would lead to very odd data container choices in the "single series" setting, such as a 3D array with two dimensions being 1 for univariate time series, or a `pd.DataFrame` with a `MultiIndex` whose highest level always has the same value. It wouldn't upset people who work with `Panel`-s, but it would probably confuse the heck out of people working with single series, such as in forecasting. In my opinion at least. If you follow this argument, you end up with different data/mtype representations for `Series` and `Panel`. Various other, perhaps more historical than principled requirements, lead to multiple mtype per scitype (e.g., eight `Panel` representations in use).

>I see the benefits of both sides: I love the idea, that the original input mtype remains (is back-transformed) for the output. From my practicle point of view, I see the downside of back-transformation in terms of computational speed and RAM for big datasets. In a perfect world, this may be a decision of the user, either by an init parameter of the transformer or a _tag = {keep_mtype = True} with default value True i.e. original mtype is returned.

Yes, indeed that's precisely the internal discussion we're having, and the two main arguments for/against.
Some are in favour of giving the user the flexibility (plus perhaps adding the `keep_mtype` flag), others are in favour of "unifying everything" (output mtypes mostly).

>I know the downsides of (multi-)variate TS with unequal length (especially in terms of memory allocation and computational speed) - but in my field, they are common practice as datasets would get too big if all variables are resampled to the same length

If that's what your data is (i.e., unequal length), then that's what your data is...
Unfortunately, python provides no good representation of a "data frame" type. A good step in the direction is `awkward-array`, have you seen that? Benefit of the mtype formalism is that we can just add another `awkward-pandas` based one or similar once that exists.

>Sometimes I wonder what contradicts simple 2D-list sturctures (i.e. [ [pd.Series(), pd.Series(), ...], [pd.Series(), pd.Series(), ...], ...], with instances as first dim and variables as second dim, comparable to df-list) as python lists are quite competend in managing memory allocation issues...

Well, we've had data container discussions for about 2 years until we decided it's a horrible time sink and it may be better to support multiple. The problem with the list-of-lists structure, if I remember, is the difficulty to index by row, to sub-set by row, selections across columns, and the lack of column names.

I also suspect all core-devs agree that `nested_univ` feels really wrong...
Might be worth do do some internal refactoring and move it to something much more efficient, now that we have the "outer" layer.

**Reply by fkiraly** [reply]
Regarding tests, there is already a suite of tests but it may be a bit gappy.
It's in the `transformations/tests` folder, and it will be run on any transformers found in `sktime`, if you are using a local set-up with `pytest`.

No single function `test_my_transformer` that you call on the class, but this is quite close.

**Reply by fkiraly** [reply]
> We still have some dtype-checking problems with the new BaseTransformer.fit / transform

what are the error messages?
You can drill down by calling `check_is` on your container with the mtype you *think* it is, then it should give you an informative error message why it's not that mtype (e.g., you named columns or indices wrong, or time points are not in monotonous order, etc).
The `datatypes` tutorial has the full specifications.

**Comment by fkiraly** [comment]
How's it going?
Let us know if you need any advice!
The transformer extension template is new so might still be a bit wonky.

**Reply by MrPr3ntice** [reply]
Some news here: We created a draft PR some days ago and passed all checks so far. Now we are looking forward to any feedback and review.

**Reply by fkiraly** [reply]
excellent, thanks!

**Reply by fkiraly** [reply]
(would you mind closing this discussion then since we agreed it´s a great idea?)

**Reply by fkiraly** [reply]
(you need to mark sth as the answer)
